pkg/translate/translateValue.go

```diff
@@ -0,0 +1,345 @@
+		enabled := name.IsComponentEnabledFromValue(vp, valueSpec)
+		// set feature enablement
+		if fe == nil || fe.outPath == "" {
+			continue
+		}
+		newP := firstCharToLowerPath(fe.outPath)
+		// Value.yaml component to IstioFeature is N:1, so if the feature is enabled by other component already, skip setting
+		curEnabled, found, _ := name.GetFromValuePath(root, newP)
+		if !found || curEnabled == false {
+			if err := setTree(root, newP, enabled); err != nil {
+				return err
+			}
+		}
+		// set component enablement
+		ce := t.ComponentEnablementMapping[vp]
+		if ce == nil || ce.outPath == "" {
+			continue
+		}
+		outP := firstCharToLowerPath(ce.outPath)
+		if err := setTreeFromValue(root, outP, enabled); err != nil {
+			return err
+		}
+	}
+
+	for vp, nsList := range t.NamespaceMapping {
+		namespace := name.NamespaceFromValue(vp, valueSpec)
+		for _, ns := range nsList {
+			if err := setTreeFromValue(root, util.PathFromString(ns), namespace); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+//translateTree is internal method for translating value.yaml tree
+func (t *ValueYAMLTranslator) translateTree(valueTree map[string]interface{},
+	cpSpecTree map[string]interface{}, path util.Path, mapping map[string]*Translation) error {
+	// translate input valueTree
+	for key, val := range valueTree {
+		newPath := append(path, key)
+		// leaf
+		if val == nil {
+			err := t.insertLeaf(cpSpecTree, newPath, val, mapping)
+			if err != nil {
+				return err
+			}
+		} else {
+			switch node := val.(type) {
+			case map[string]interface{}:
+				err := t.translateTree(node, cpSpecTree, newPath, mapping)
+				if err != nil {
+					return err
+				}
+			case []interface{}:
+				for _, newNode := range node {
+					newMap, ok := newNode.(map[string]interface{})
+					if !ok {
+						return fmt.Errorf("fail to translate slice")
+					}
+					err := t.translateTree(newMap, cpSpecTree, newPath, mapping)
+					if err != nil {
+						return err
+					}
+				}
+			// leaf
+			default:
+				err := t.insertLeaf(cpSpecTree, newPath, val, mapping)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (t *ValueYAMLTranslator) insertLeaf(root map[string]interface{}, newPath util.Path,
+	val interface{}, mapping map[string]*Translation) (errs util.Errors) {
+	// Must be a scalar leaf. See if we have a mapping.
+	valuesPath, m := getValuesPathMapping(mapping, newPath)
+	switch {
+	case m == nil:
+		break
+	case m.translationFunc == nil:
+		// Use default translation which just maps to a different part of the tree.
+		errs = util.AppendErr(errs, defaultTranslationFunc(m, root, valuesPath, val))
+	default:
+		// Use a custom translation function.
+		errs = util.AppendErr(errs, m.translationFunc(m, root, valuesPath, val))
+	}
+	return errs
+}
+
+// componentString renders a template of the form <path>{{.ComponentName}}<path> with
+// the supplied parameters.
+func componentString(tmpl string, componentName string) string {
```

Comment:
it would be more meaningful to call this something like "renderWithComponentName"